### ISSUE TITLE
403 Error

### ISSUE THREAD

**Issue by kyleplo** [towerofnix/scratch-notifs#3]
When I go to https://scratch.mit.edu/session/, I get a 403 error.

Also, when I did the code below in the console of the Scratch homepage, I got "{<2 line breaks>}"
```javascript
fetch("https://scratch.mit.edu/session/").then(function (r){return r.text()}).then(function (r){console.log(r)})
```

**Comment by towerofnix** [comment]
Yeah, just fetch-ing your session info like that won't work, because fetch won't send any personally-identifying data -- that is, your cookies. But even sending your cookies correctly (which is what opening the page in your browser does) also has the 403 error, as you see.

I did [a quick comparison](https://gist.github.com/towerofnix/5cd07a8ee65af870dd1fbf1485c3633d/revisions) between the request that scratch.mit.edu makes and what `fetch` makes, and it looks like the issue is that in order for it to work, there needs to be an `X-Requested-With: XMLHttpRequest` header. (I tested this with [a modified fetch](https://gist.github.com/towerofnix/b9450dc13aeb51cfaa945298e66638cc), which worked.)

This is a bit of a pain, because my code currently asks people to manually go and grab their session ID... and now there's no easy way, since opening the page in your browser doesn't work anymore. I could ask them to copy-paste some JavaScript code into their javascript console, but that reeks of "I'm trying to hack you!!!!"... although I'm not sure asking for someone's session ID in the first place is any better.

Although, I'm glad that it's now a 403 when you just open the page in the browser. After all, there *could* be actually-malicious people out there who want to steal a session ID from an unsuspecting user/rival, so this prevents that from being nearly as simple to do. (Before you could just go, "Hi, can you please open this page and copy-paste what's in it here? That will help me fix this bug." But not anymore, obviously.) It also helps with people who are, say, doing a screen-recording; if someone making a video opens /session/ after being requested by a not-malicious program such as mine, then they might inadvertently publish their session ID to the whole world (by way of sharing the video). That wouldn't be any good, either.

Anyways.. thanks for bringing this up! I'm not sure how to fix scratch-notifs right now, but I'll definitely make a note of this in scratch-api-unofficial-docs. - Ah, edit, looks like this was already done and merged, courtesy of @joker314 - https://github.com/towerofnix/scratch-api-unofficial-docs/pull/26. (Heya jbs, you might have some input here? :P)

**Comment by joker314** [comment]
This might've been written before I had GitHub notifications set up, so sorry for the late response!

It's now _even harder_ to get someone to copy-paste something into the console, because if they didn't yet know the risks, Scratch will scream at them with a warning message llk/scratch-www#1899.

> # Stop!
>
> This is part of your browser intended for developers. If someone told you to copy-and-paste something here, don't do it! It could allow them to take over your Scratch account, delete all of your projects, or do many other harmful things. If you don't understand what exactly you are doing here, you should close this window without doing anything.

Yet the request has to come from Scratch's origin or else <abbr title="Same-Origin Policy">SOP</abbr>

# My 💡 idea:

Make the request anyway, and get the user to look for the response data in the "Network" tab of the console. That doesn't work, though, so we need to think of a better way.

We could get them to give us their cookie, but that feels morally wrong.

We could tell them they don't need to give us the token (but we need to think about what happens when, one day, they will).

Finally, we could petition Scratch to add some kind of "OAuth" (but with better security) to Scratch. *shrugs*

**Comment by towerofnix** [comment]
Right. I'm not sure what the best option here is. Btw,

> Make the request anyway, and get the user to look for the response data in the "Network" tab of the console. 

Instead of doing that, we can ask the user to open up their messages page, then find the request to api.scratch.mit.edu, then copy the X-token request header. The problem is, while that's reliable, it's not *remotely* practical.

For myself testing things with the API, I've added [a thing](https://github.com/towerofnix/scratch-client-omg/commit/4414ddfb0bba48a2bee40c5ebe6616caa0b93592) to scratch-client-omg to print my authentication token. But it isn't really good to ask users to download and run my program first (and that's hardly user-friendly to anyone who doesn't use node).

**Comment by joker314** [comment]
LLK/scratch-rest-api#1 may be relevant -- the suggestion of adding an OAuth system.

**Comment by joker314** [comment]
How about adding a page where multiple methods are given, and users can choose the one they want

- Finding the token in their Network tab
- Pasting a Totally Safe console code
- Downloading a Node program
- Giving us their Scratch cookie and us somehow converting it (using a server, perhaps)

**Comment by towerofnix** [comment]
> Giving us their Scratch cookie and us somehow converting it (using a server, perhaps)

No. The absolute baseline rule is that cookies/sessions/tokens will *never, ever* be sent to anywhere besides api.scratch.mit.edu.

> Pasting a Totally Safe console code, Downloading a Node program

Neither of these are or look good. Downloading a node program means running arbitrary code on your computer - plus, the program requires you enter your username and password, which would be even more suspicious. And pasting a console code is icky.

> Finding the token in their Network tab

This is not functionally any different than asking the user to run a code in their console because it still gives scratch-notifs full access to the user's account (or rather, everything that requires authentication on api.scratch.mit.edu). But, of course, that's already what we're doing, with your trust that we'll only use it for checking your message count.

To be clear, I want you to know we're discussing this in the *hypothetical*. I've likely given up work on scratch-notifs. If I ever create something similar, it'll probably be built into scratch-client-omg.
